### PR TITLE
SW-8291 Implemented updating plot species count and t0 data 

### PR DESCRIPTION
**Pull request by tommylau523**
This PR primarily handles recalculation triggers

Co-authored by Claude, for debugging, tracing edge cases, and building out test scenarios

### CODE REVIEW

**Comment by tommylau523** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4333?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#4363** <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4363?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4362** <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4362?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4333** <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4333?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4333?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#4332** <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4332?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4343** <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4343?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4331** <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4331?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4330** <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4330?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4329** <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4329?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4328** <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4328?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#4342** <a href="https://app.graphite.com/github/pr/terraware/terraware-server/4342?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>merge-queue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/terraware?ref=merge-queue-instructions-in-stack-comment&prId=4333).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/main/kotlin/com/terraformation/backend/tracking/util/ObservationResultsScope.kt [hunk 1]
```diff
@@ -52,9 +54,18 @@ interface ObservationResultsScope<ID : Any, HistoryId : Any> :
...
 
-  fun anyChildHasNullSurvivalRateCondition(observationId: ObservationId): Condition
+  /**
+   * Condition that covers all result table rows that could be affected by a change to the scoped
+   * entity. For ID-based scopes this is the same as [observedTotalsCondition]. For plotId-based
+   * scopes it expands to include all observations for every substratum/stratum the plot has ever
+   * historically belonged to, ensuring that observations which reference the plot's data via
+   * [observationIdForPlot] are also updated.
+   */
+  val survivalRatesCondition: Condition
```
Comment (tommylau523) [inline src/main/kotlin/com/terraformation/backend/tracking/util/ObservationResultsScope.kt]: This should probably be renamed to survivalRateRecalculationCondition.

File: src/main/kotlin/com/terraformation/backend/tracking/db/ObservationStore.kt [hunk 2]
```diff
@@ -1883,13 +1893,102 @@ class ObservationStore(
...
   private fun recalculateSurvivalRateResults(
       observationId: ObservationId,
       plantingSiteId: PlantingSiteId,
   ) {
     // Update tables with latest total values
     updateObservationResults(observationId, plantingSiteId)
 
+    val plotIds =
+        dslContext
+            .selectDistinct(OBSERVATION_PLOT_RESULTS.MONITORING_PLOT_ID)
```
Comment (sgrimm) [inline src/main/kotlin/com/terraformation/backend/tracking/db/ObservationStore.kt]: Why is the `selectDistinct` needed here? There can't be more than one result for a single monitoring plot in a single observation, can there?

File: src/main/kotlin/com/terraformation/backend/tracking/util/ObservationResultsScope.kt [hunk 3]
```diff
@@ -276,6 +328,20 @@ class ObservationResultsStratum(
   override val observedTotalsCondition =
       OBSERVATION_STRATUM_RESULTS.STRATUM_HISTORY_ID.`in`(stratumHistorySelect)
 
+  override val survivalRatesCondition: Condition
```
Comment (tommylau523) [inline src/main/kotlin/com/terraformation/backend/tracking/util/ObservationResultsScope.kt]: This is likely an "over-update", because we are recalculating for basically every stratum that has ever contained this plot.

In the followup PRs, we will introduce a table of observation dependencies, which will help with reducing the update scope to a set of observations.

**Review by sgrimm — APPROVED** [review]
